SecureRandom, Hashing, HMAC, Signature, SymmetricEncryption

emarx opened the pull request:
Hi! Coincindentally, I started a `zio-crypto` project this weekend, too. I've implemented `SecureRandom`, `Hashing`, and `HMAC` so far, with plans to implement `Signature`, `JWT`, `PasswordHashing`, `SymmetricEncryption`, `AsymmetricEncryption`, and `HybridEncryption`. Is this the same motivation as this project? If so, hopefully we can merge this in to give you a head start! Comes with a website for all implemented features and nice documentation in the code.

emarx commented:
Looks like we used some slightly different templates -- happy to clean this up if the goals are the same, but wanted to get the PR up quickly to have the discussion

jdegoes commented:
@emarx Would love to have your contributions. What template did you use?

mijicd commented:
@jdegoes from what i see, the biggest differences are "resurrected" CircleCI and  dockerized docs build. There are some differences in scala versions but that's an easy fix.

@emarx you should take the actions setup from one of the working projects (e.g. zio-actors, zio-prelude), and reach out to me or @softinio if you need any help with that. This can be tackled in a separate PR.

emarx commented:
@jdegoes I used https://github.com/zio/zio-project-seed.g8 to set up the project.

@mijicd happy to take the boilerplate in the existing repo and close this PR if that makes sense!

mijicd commented:
You can wrap up the functionality here, we can port the CI config later. And we should definitely update the project seed :)

khajavi commented:
> Hi! Coincindentally, I started a `zio-crypto` project this weekend, too. I've implemented `SecureRandom`, `Hashing`, and `HMAC` so far, with plans to implement `Signature`, `JWT`, `PasswordHashing`, `SymmetricEncryption`, `AsymmetricEncryption`, and `HybridEncryption`. Is this the same motivation as this project? If so, hopefully we can merge this in to give you a head start! Comes with a website for all implemented features and nice documentation in the code.

Hi @emarx 
Great to see your work. I'm currently planning to work on hashing algorithms, digital signatures, and also PRNG, so it seems we have the same motivation.

emarx commented:
@jdegoes @mijicd -- I've removed the template changes, and left only the new services and their documentation / tests. I think this is in a reasonable state to merge

mijicd commented:
@emarx build related changes are in place, @jdegoes can sign off the rest. Thanks for taking care of this!

khajavi commented on zio-crypto/shared/src/main/scala/zio/crypto/signature/Signature.scala:
```diff
@@ -0,0 +1,93 @@
+  val live: ULayer[Signature] = ZLayer.succeed(new Service {
+
+    def genKey(alg: SignatureAlgorithm): Task[SignatureKeyPair] = Task.effect {
+      val keyPairGenerator = KeyPairGenerator.getInstance("EC")
```
I think setting the key size with the `initialize` method is forgotten here based on ECDSASHA256 or ECDSASHA384 or ECDSASHA512. Am I right?

emarx replied:
I don't believe so. The algorithm is specified on line 46 in this file.

khajavi replied:
Yes, that's right.

khajavi reviewed (commented):
Hey @emarx 
I reviewed your work and add some comments.

emarx commented on zio-crypto/shared/src/main/scala/zio/crypto/signature/Signature.scala:
```diff
@@ -0,0 +1,93 @@
+}
+
+case class SignaturePrivateKey(key: PrivateKey, algorithm: SignatureAlgorithm)
+case class SignaturePublicKey(key: PublicKey, algorithm: SignatureAlgorithm)
```
That was the goal of these types (`SignaturePrivateKey`, `SignaturePublicKey` and `SignatureKeyPair`.) If we want other keys, these types can be made fancier without changing the interface.

khajavi replied:
But these types don't cover this goal. as they depend on `java.security.PrivateKey`.

emarx commented on zio-crypto/shared/src/main/scala/zio/crypto/hashing/Hashing.scala:
```diff
@@ -0,0 +1,115 @@
+  type Hashing = Has[Hashing.Service]
+
+  trait Service {
+    def hash(m: Seq[Byte], alg: HashAlgorithm): Task[MessageDigest[Seq[Byte]]]
```
Array works for me! I pushed up a commit with all `Array` instead of `Seq`

emarx commented:
Thanks for your review, @khajavi! Comments addressed and one new commit pushed

jdegoes commented on website/yarn.lock:
```diff
@@ -0,0 +1,6674 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
This file should not be committed.

jdegoes commented on zio-crypto/shared/src/main/scala/zio/crypto/hashing/Hashing.scala:
```diff
@@ -0,0 +1,115 @@
+import java.nio.charset.Charset
+import java.security.{ MessageDigest => JMessageDigest }
+
+sealed class HashAlgorithm(val name: String)
```
`sealed trait`

jdegoes commented on zio-crypto/shared/src/main/scala/zio/crypto/hashing/Hashing.scala:
```diff
@@ -0,0 +1,115 @@
+package zio.crypto.hashing
```
`zio.crypto.hash` (shorter)

jdegoes commented on zio-crypto/shared/src/main/scala/zio/crypto/hashing/Hashing.scala:
```diff
@@ -0,0 +1,115 @@
+  case object SHA512 extends HashAlgorithm("SHA-512")
+
+  // DO NOT USE IN A SECURE CONTEXT
+  case object MD5 extends HashAlgorithm("MD5")
```
Let's solve this in a different way:

```scala
trait Secure[-Algorithm]
```

Then inside HashAlgorithm companion object:

```scala
implicit val secureSha256: Secure[SHA256.type] = new Secure[SHA256.type]{}
...
```

Then make a helper in `zio.crypto`:

```scala
def unsecure[A](f: Secure[Any] => A): A = f(new Unsecure[Any]{})
```

Then any code which requires a `HashAlgorithm` should also require `Secure` for that hash algorithm.

e.g.:

```scala
def hash[Alg <: HashAlgorithm](m: Array[Byte], alg: HashAlgorithm)(implicit s: Secure[HashAlgorihm]): Task[MessageDigest[Array[Byte]]]
```

Now if you want to use a secure algorithm, it will be available. But if you want to use an insecure algorithm, you have to do:

```scala
unsecure { implicit s =>
 ...
}
```

The technique can be used for anything, even if it's not a hash.

jdegoes commented on zio-crypto/shared/src/main/scala/zio/crypto/hashing/Hashing.scala:
```diff
@@ -0,0 +1,115 @@
+  type Hashing = Has[Hashing.Service]
+
+  trait Service {
+    def hash(m: Array[Byte], alg: HashAlgorithm): Task[MessageDigest[Array[Byte]]]
```
`Chunk[Byte]` should always be used instead of `Array[Byte]`. `Array` is mutable, but `Chunk` is not, and Chunk does not box bytes or other primitives, like Array (it's based on Array).

jdegoes commented on zio-crypto/shared/src/main/scala/zio/crypto/hashing/Hashing.scala:
```diff
@@ -0,0 +1,115 @@
+
+  trait Service {
+    def hash(m: Array[Byte], alg: HashAlgorithm): Task[MessageDigest[Array[Byte]]]
+    def verify(m: Array[Byte], digest: MessageDigest[Array[Byte]], alg: HashAlgorithm): Task[Boolean]
```
If a function is deterministic and free of observable side-effects, then it should not return the value wrapped in `Task`. 

As far as I can see, all of these can be pure functions. Right?

emarx replied:
`hash` can throw `NoSuchAlgorithmException`, which is why I had wrapped it in `Task`. `verify` calls `hash` on `m` to check against `digest`, which is why `verify` is a `Task` as well. Perhaps that is an acceptable and unlikely failure? The remaining possible failure is in the `String` versions of these functions, where the supplied base 64 encoded string may be invalid.

emarx commented on zio-crypto/shared/src/test/scala/zio/crypto/hash/HashSpec.scala:
```diff
@@ -0,0 +1,74 @@
+
+object HashSpec extends DefaultRunnableSpec {
+
+  private val assertCompletesM = assertM(UIO(true))(isTrue)
```
PR here adding this to `zio/zio`: https://github.com/zio/zio/pull/4740

jdegoes commented on zio-crypto/shared/src/main/scala/zio/crypto/mac/HMAC.scala:
```diff
@@ -0,0 +1,163 @@
+import javax.crypto.spec.SecretKeySpec
+import javax.crypto.{ KeyGenerator, Mac, SecretKey }
+
+sealed class HMACAlgorithm(val name: String)
```
`sealed trait`

jdegoes commented on zio-crypto/shared/src/main/scala/zio/crypto/mac/HMAC.scala:
```diff
@@ -0,0 +1,163 @@
+  case object HMACSHA512 extends HMACAlgorithm("HmacSHA512")
+}
+
+case class HMACSerializedKey(value: String)     extends AnyVal
```
`final case class`

jdegoes commented on zio-crypto/shared/src/main/scala/zio/crypto/mac/HMAC.scala:
```diff
@@ -0,0 +1,163 @@
+  type HMAC = Has[HMAC.Service]
+
+  trait Service {
+    def sign(m: Array[Byte], k: HMACSecretKey): Task[HMACObject[Array[Byte]]]
```
Pure functions? Also Chunk, etc.

emarx replied:
`java.crypto.Mac.init`, which is used in the body of `sign` throws if the key is invalid. `Chunk` addressed here and elsewhere :+1:

emarx commented:
@jdegoes I believe I have addressed your comments, including in removing `Task` wrappers. However, in `HMAC` and `Hash`, please take a look at the comments that say `"May throw <X>"`. These methods were wrapped in `Task` because of these potential exceptions. `Signature`, any key generation, and `SymmetricEncryption` are left with `Task` wrappers because they exhibit non-determinism.

jdegoes commented on zio-crypto/shared/src/main/scala/zio/crypto/random/SecureRandom.scala:
```diff
@@ -0,0 +1,107 @@
+    def nextBytes(length: Int): Task[Array[Byte]]
+    def nextString(entropyBytes: Int): Task[String]
+    def setSeed(seed: Long): UIO[Unit]
+    def getJavaSecureRandom: UIO[JSecureRandom]
```
Another pattern that libraries have used:

```scala
def execute[A](f: JSecureRandom => A): ZIO[Any, Throwable, A]
```

This way it's harder to "leak" the mutable interface.

jdegoes commented on zio-crypto/shared/src/main/scala/zio/crypto/symmetric/SymmetricEncryption.scala:
```diff
@@ -0,0 +1,120 @@
+    val NISTIvLengthBytes = 12
+
+    private def getInstance: Task[Cipher] =
+      Task.effect(Cipher.getInstance("AES/GCM/NoPadding"))
```
This will not be cached. Every time `getInstance` is sequenced into a computation, it will be called again. Does this have potential to be a performance problem?

emarx replied:
My understanding is that `Cipher` is a mutable object that you use for a single signature, and that using it more than once would result in an error.
